### DISCUSSION TITLE
What was the reasoning behind choosing to use webOb?

### DISCUSSION THREAD

**Discussion by michaeltoohig** [StackStorm/st2#5820] [General]
Just curious why webOb is the selected library for handling web requests? I've not seen it before and when I needed something that was not too opinionated I would reach for Flask. From what I'm learning webOb seems to be a step even lower than Flask in the opinionated framework spectrum or actually more like werkzeug which Flask is built on top of.

I would like to have it know where I can read some code where webOb afforded some special usecase in the stackstorm codebase where other frameworks have gotten in the way. Mostly because I'm interested to learn something new and appreciate how I may want to use webOb in my future too.

Or perhaps there is no special reason webOb was used other than early design choice and any framework would have been just fine?

Thank you.

**Comment by arm4b** [comment]
I could find the history in this PR: https://github.com/StackStorm/st2/pull/2727 from the 2017 when the `webob` lib was first imported.

From that, some reasons were:

> Instead of monolithic framework, I propose to use a number of smaller single purpose libraries and build our router and controller around them. In this case, I've replaced pecan with three libraries:
> - WebOb provides a number of user-friendly utility functions to work with HTTP requests and responses. For example, it allows you to access json-encoded request data using req.json() instead of parsing it manually from WSGI environ object. It also allows you to construct the response and only call start_response at the last possible moment.
> - Routes handles URL matching and templating allowing us to comply with OpenAPI spec regarding path templating without diving deep into regex replacements.
> - swagger-spec-validator handles spec validation and helps us spot an error in our spec during initialization


More:

> As you see, I've tried flask and connexion first but figured you can't redefine error reporting in this case.

and more:

> As for being battle tested, both WebOb and routes are used by reddit and are also a parts of other frameworks (pecan uses webob and routes is a part of pylons) so they are getting plenty.

... and the whole discussion in the [REST Services v2 #2727 (2017) ](https://github.com/StackStorm/st2/pull/2727)

Surprised it was possible to find it and love how we have a meaningful commit history, PR discussions, and everything is so detailed.

Hope that helps ;)
